FFT: disable dimension wrapping for scalar tensors

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #89234

Fixes #88985

By default, `maybe_wrap_dim` allows through `dim=0` or `dim=-1`
for scalar tensors which leads to an invalid dimension being used to
index into `tensor.sizes()` as in the code sample from the issue.

cc @mruberry

peterbell10 committed:
FFT: disable dimension wrapping for scalar tensors

Fixes #88985

By default, `maybe_wrap_dim` allows through `dim=0` or `dim=-1`
for scalar tensors which leads to an invalid dimension being used to
index into `tensor.sizes()` as an example.

[ghstack-poisoned]

mruberry commented:
@peterbell10 could you elaborate on this change with a before and after examples? Are there any backwards compatibility concerns here?

peterbell10 commented:
Take `fft` as an example. Before this PR I get this
```
In [1]: import torch
   ...: a = torch.rand(())
   ...: torch.fft.fft(a)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Input In [1], in <cell line: 3>()
      1 import torch
      2 a = torch.rand(())
----> 3 torch.fft.fft(a)

RuntimeError: Invalid number of data points (0) specified
```

`fft` must transform exactly one dimension, but here it's allowed `dim=-1` (the default for `fft`) on a zero-dimensional tensor. The error in this case makes no sense  because it's reading from uninitialized memory that happens to be a zero but in general might be garbage.

After this change, the error message becomes 
```
IndexError: Dimension specified as -1 but tensor has no dimensions
```

There is no backward compatibility implication because the previous zero-dim behavior depended on undefined behavior and was purely a bug.

mruberry commented on aten/src/ATen/WrapDimUtils.h:
```diff
@@ -43,9 +43,19 @@ inline int64_t maybe_wrap_dim(
     int64_t ndims,
-    int64_t dim_post_expr) {
+    int64_t dim_post_expr,
+    bool wrap_scalars = true) {
```
Let's extend the comment describing this function to include a description of each parameter and what they do

mruberry commented on aten/src/ATen/WrapDimUtils.h:
```diff
@@ -70,8 +80,12 @@ inline void maybe_wrap_dims_n(
 template <typename Container>
-inline void maybe_wrap_dims(Container& dims, int64_t dim_post_expr) {
-  return maybe_wrap_dims_n(dims.data(), dims.size(), dim_post_expr);
+inline void maybe_wrap_dims(
```
Improved comment needed here, too

mruberry commented on torch/testing/_internal/opinfo/definitions/fft.py:
```diff
@@ -65,6 +67,24 @@ def __init__(
 
 
+def error_inputs_fft(op_info, device, **kwargs):
+    make_arg = partial(make_tensor, device=device, dtype=torch.float32)
```
Add a comment describing the error tested for both these `ErrorInput`s

mruberry reviewed (approved):
Cool! But see requests for some additional comments

peterbell10 committed:
Update on "FFT: disable dimension wrapping for scalar tensors"


Fixes #88985

By default, `maybe_wrap_dim` allows through `dim=0` or `dim=-1`
for scalar tensors which leads to an invalid dimension being used to
index into `tensor.sizes()` as in the code sample from the issue.

cc mruberry

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge
